File: CS/E2979MVC/Controllers/HomeController.cs
using E2979MVC.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace E2979MVC.Controllers
{
    public class HomeController : Controller
    {
        //
        // GET: /Home/
        NorthwindDataContext context = new NorthwindDataContext();
        public ActionResult Index()
        {
            return View();
        }

        public ActionResult ProductsGridViewPartial()
        {
            var model = context.Products;
            ViewData["Categories"] = context.Categories;
            return PartialView(model);
        }

        public ActionResult GridLookupPartial(int? CurrentCategory)
        {
            ViewData["Categories"] = context.Categories;
            if (CurrentCategory == null)
                CurrentCategory = -1;
            return PartialView(new Product() { CategoryID = (int)CurrentCategory });
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult ProductsGridViewAddNewPartial(Product item)
        {
            var model = context.Products;
            ViewData["Categories"] = context.Categories;
            if (ModelState.IsValid)
            {
                try
                {
                    model.InsertOnSubmit(item);
                    context.SubmitChanges(); //uncomment this line to save changes to a data base
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";
           
            return PartialView("_ProductsGridViewPartial", model);
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult ProductsGridViewUpdatePartial(Product item)
        {
            var model = context.Products;
            ViewData["Categories"] = context.Categories;
            if (ModelState.IsValid)
            {
                try
                {
                    var modelItem = model.FirstOrDefault(it => it.ProductID == item.ProductID);
                    if (modelItem != null)
                    {
                        modelItem.ProductName = item.ProductName;
                        modelItem.UnitPrice = item.UnitPrice;
                        modelItem.CategoryID = item.CategoryID;
                        modelItem.QuantityPerUnit = item.QuantityPerUnit;                      
                     //   context.SubmitChanges(); uncomment this line to save changes to a data base
                    }
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";
           
            return PartialView("ProductsGridViewPartial", model);
        }
        [HttpPost, ValidateInput(false)]
        public ActionResult ProductsGridViewDeletePartial(System.Int32 ProductID)
        {

            var model = context.Products;
            ViewData["Categories"] = context.Categories;
            if (ProductID >= 0)
            {
                try
                {
                    var item = model.FirstOrDefault(it => it.ProductID == ProductID);
                    if (item != null)
                        model.DeleteOnSubmit(item);
                    // context.SubmitChanges(); uncomment this line to save changes to a data base
                }
                catch (Exception e)
                {
                    ViewData["EditError"] = e.Message;
                }
            }

            return PartialView("_ProductsGridViewPartial", model);
        }
    }
}

File: VB/E2979MVC/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports E2979DVC.Models
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc

Namespace E2979DVC.Controllers
	Public Class HomeController
		Inherits Controller
		'
		' GET: /Home/
		Private context As New NorthwindDataContext()
		Public Function Index() As ActionResult
			Return View()
		End Function

		Public Function ProductsGridViewPartial() As ActionResult
			Dim model = context.Products
			ViewData("Categories") = context.Categories
			Return PartialView(model)
		End Function

		Public Function GridLookupPartial(ByVal CurrentCategory? As Integer) As ActionResult
			ViewData("Categories") = context.Categories
			If Not CurrentCategory.HasValue Then
				CurrentCategory = -1
			End If
			Return PartialView(New Product() With {.CategoryID = CInt(Fix(CurrentCategory))})
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function ProductsGridViewAddNewPartial(ByVal item As Product) As ActionResult
			Dim model = context.Products
			ViewData("Categories") = context.Categories
			If ModelState.IsValid Then
				Try
					model.InsertOnSubmit(item)
					context.SubmitChanges() 'uncomment this line to save changes to a data base
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If

			Return PartialView("_ProductsGridViewPartial", model)
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function ProductsGridViewUpdatePartial(ByVal item As Product) As ActionResult
			Dim model = context.Products
			ViewData("Categories") = context.Categories
			If ModelState.IsValid Then
				Try
					Dim modelItem = model.FirstOrDefault(Function(it) it.ProductID = item.ProductID)
					If modelItem IsNot Nothing Then
						modelItem.ProductName = item.ProductName
						modelItem.UnitPrice = item.UnitPrice
						modelItem.CategoryID = item.CategoryID
						modelItem.QuantityPerUnit = item.QuantityPerUnit
					 '   context.SubmitChanges(); uncomment this line to save changes to a data base
					End If
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			Else
				ViewData("EditError") = "Please, correct all errors."
			End If

			Return PartialView("ProductsGridViewPartial", model)
		End Function
		<HttpPost, ValidateInput(False)> _
		Public Function ProductsGridViewDeletePartial(ByVal ProductID As System.Int32) As ActionResult

			Dim model = context.Products
			ViewData("Categories") = context.Categories
			If ProductID >= 0 Then
				Try
					Dim item = model.FirstOrDefault(Function(it) it.ProductID = ProductID)
					If item IsNot Nothing Then
						model.DeleteOnSubmit(item)
					End If
					' context.SubmitChanges(); uncomment this line to save changes to a data base
				Catch e As Exception
					ViewData("EditError") = e.Message
				End Try
			End If

			Return PartialView("_ProductsGridViewPartial", model)
		End Function
	End Class
End Namespace